pkg/apis/ark/v1/volume_snapshot.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2018 the Heptio Ark contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeSnapshot is... TODO
+type VolumeSnapshot struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   VolumeSnapshotSpec   `json:"spec"`
+	Status VolumeSnapshotStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeSnapshotList is a list of VolumeSnapshots.
+type VolumeSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []VolumeSnapshot `json:"items"`
+}
+
+// VolumeSnapshotSpec defines the specification for an Ark VolumeSnapshot.
+type VolumeSnapshotSpec struct {
+	// Type is the type of the disk/volume in the cloud provider
+	// API.
+	Type string `json:"type"`
+
+	// AvailabilityZone is the where the volume is provisioned
+	// in the cloud provider.
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
+
+	// Iops is the optional value of provisioned IOPS for the
+	// disk/volume in the cloud provider API.
+	Iops *int64 `json:"iops,omitempty"`
+
+	// Backup is a string containing the name of name of the Ark backup this snapshot is associated with.
+	Backup string `json:"backup"`
+
+	// Location is a string containing the name of a VolumeSnapshotLocation.
+	Location string `json:"volumeLocation"`
+}
+
+// VolumeSnapshotStatus captures the current status of an Ark VolumeSnapshot.
+type VolumeSnapshotStatus struct {
```

Comment:
We'll also need an ID field for a UUID generated by Ark.

This is because the current `SnapshotID` applies to a snapshot inside of a region. If we copy the snapshot to a new region, the cloud provider will assign it a new `SnapshotID`, even though the contents are the same. Using a UUID we generate allows us to view them as the same logical snapshot from Ark's perspective.

See [this note](https://docs.google.com/document/d/1vGz53OVAPynrgi5sF0xSfKKr32NogQP-xgXA1PB6xMc/edit#bookmark=id.br0hvcpuqy0n) and [this one](https://docs.google.com/document/d/1vGz53OVAPynrgi5sF0xSfKKr32NogQP-xgXA1PB6xMc/edit#bookmark=id.j9gluvt46fqv) in the design doc.

Reply:
I like `ProviderSnapshotID`. I don't think we actually need the UUID until we get to replication in the next release, but we can add it and leave it unused for now

Reply:
I would think we'd use it with #804, too, which is part of v0.10.0